rust/src/io/exec/scan.rs

```diff
@@ -56,70 +93,65 @@ impl LanceStream {
         projection: Arc<Schema>,
         read_size: usize,
         batch_readahead: usize,
+        fragment_readahead: usize,
         with_row_id: bool,
+        scan_in_order: bool,
     ) -> Result<Self> {
-        let (tx, rx) = mpsc::channel(batch_readahead);
-
         let project_schema = projection.clone();
-        let io_thread = tokio::spawn(async move {
-            'outer: for frag in fragments.as_ref() {
-                if tx.is_closed() {
-                    return;
-                }
-                let file_fragment = FileFragment::new(dataset.clone(), frag.clone());
-                let reader = match file_fragment.open(project_schema.as_ref()).await {
-                    Ok(mut r) => {
-                        if with_row_id {
-                            r.with_row_id();
-                        };
-                        r
-                    }
-                    Err(e) => {
-                        tx.send(Err(DataFusionError::from(e)))
-                            .await
-                            .expect("Scanner sending error message");
-                        // Stop reading.
-                        break;
-                    }
-                };
-                for batch_id in 0..reader.num_batches() {
-                    let rows_in_batch = reader.num_rows_in_batch(batch_id);
-                    for start in (0..rows_in_batch).step_by(read_size) {
-                        let result = reader
-                            .read_batch(batch_id, start..min(start + read_size, rows_in_batch))
-                            .await;
-                        if tx.is_closed() {
-                            // Early stop
-                            break 'outer;
-                        }
-                        if let Err(err) = tx.send(result.map_err(|e| e.into())).await {
-                            eprintln!("Failed to scan data: {err}");
-                            break 'outer;
-                        }
-                    }
-                }
-            }
-
-            drop(tx)
-        });
+
+        let file_fragments = fragments
+            .iter()
+            .map(|fragment| FileFragment::new(dataset.clone(), fragment.clone()))
+            .collect::<Vec<_>>();
+
+        let inner_stream = if scan_in_order {
+            stream::iter(file_fragments)
+                .then(move |file_fragment| {
+                    open_file(file_fragment, project_schema.clone(), with_row_id)
+                })
+                .map_ok(move |reader| scan_batches(reader, read_size).buffered(batch_readahead))
+                .try_flatten()
+                .boxed()
+        } else {
+            stream::iter(file_fragments)
+                .then(move |file_fragment| {
+                    open_file(file_fragment, project_schema.clone(), with_row_id)
+                })
+                .map_ok(move |reader| {
+                    scan_batches(reader, read_size).buffer_unordered(batch_readahead)
```

Comment:
So here it means that it reads `batch_readhead * fragment_readhead` in memory? 

How many actual threads are created in this case? It is not clear to me.

Might want to add some document to describe the expectation of # of threads and # of batches in flight

Reply:
> So here it means that it reads batch_readhead * fragment_readhead in memory?

I think you are right, but that's also not what we want. Took a while to figure out how to rearrange it, but I *think* my latest commit should make it so `batch_readahead` is enforced across fragments, so no matter how many fragments are read concurrently only `batch_readahead` batches will be buffered. (I put the buffer_unordered after the flatten.)

> How many actual threads are created in this case? It is not clear to me.

As I understand it, tokio will dynamically schedule them across it's thread pool, which defaults to 1 thread per core. From the [tokio docs](https://docs.rs/tokio/latest/tokio/runtime/index.html#multi-thread-scheduler):

> The multi-thread scheduler executes futures on a thread pool, using a work-stealing strategy. By default, it will start a worker thread for each CPU core available on the system.

> Might want to add some document to describe the expectation of # of threads and # of batches in flight

Agreed. This code can be a little confusing. I've added some comments.

Reply:
> but I think my latest commit should make it so batch_readahead is enforced across fragments,

I feel it is ok to make `batch_readahead` just limited to one open fragment, which could be the least surprise with the official pyarrow API ?

https://arrow.apache.org/docs/python/generated/pyarrow.dataset.Scanner.html

Reply:
it seems as desired after reading the [doc](https://docs.rs/futures/latest/futures/stream/struct.TryFlatten.html). lets merge them :)